File: cpp/62. Unique Paths.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;

/*
dp[x][y] = dp[x+1][y] + dp[x][y+1]
dp[m-1][n-1] = 1
ans = dp[0][0]
*/

class Solution {
public:
    // int uniquePaths(int m, int n) {
    //     // DP, Time: O(n^2), Space: O(n^2)
    //     int dp[m][n]; fill(&dp[0][0], &dp[0][0]+m*n, 0); dp[m-1][n-1] = 1;
    //     for(int x = m-1; x >= 0; x--) {
    //         for(int y = n-1; y >= 0; y--) {
    //             if(x==m-1 && y==n-1) continue;
    //             if(x+1>=0 && x+1<m) dp[x][y] += dp[x+1][y];
    //             if(y+1>=0 && y+1<n) dp[x][y] += dp[x][y+1];
    //         }
    //     }
    //     return dp[0][0];
    // }
    int uniquePaths(int m, int n) {
        // DP, Time: O(n^2), Space: O(n)
        int dp[n]; fill(dp, dp+n, 0); dp[n-1] = 1;
        for(int x = m-1; x >= 0; x--) {
            for(int y = n-1; y >= 0; y--) {
                // if(x+1>=0 && x+1<m) dp[x][y] += dp[x+1][y];
                // if(y+1>=0 && y+1<n) dp[x][y] += dp[x][y+1];
                if(x==m-1 && y==n-1) continue;
                if(y+1>=0 && y+1<n) dp[y] = dp[y] + dp[y+1];
            }
        }
        return dp[0];
    }
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    Solution solution; int m, n, ans;   

    solution = Solution();
    m = 3, n = 7;
    ans = solution.uniquePaths(m, n);
    assert(ans==28);

    solution = Solution();
    m = 3, n = 2;
    ans = solution.uniquePaths(m, n);
    assert(ans==3);

    solution = Solution();
    m = 7, n = 3;
    ans = solution.uniquePaths(m, n);
    assert(ans==28);

    solution = Solution();
    m = 3, n = 3;
    ans = solution.uniquePaths(m, n);
    assert(ans==6);
}

File: python/62-unique-paths.py

'''
DP: O(m*n)
'''


class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        "DP, Time: O(mn), Space: O(mn)"
        dp = [[0]*n for _ in range(m)]
        for i in range(m):
            for j in range(n):
                if i == 0 or j == 0:
                    dp[i][j] = 1
                else:
                    dp[i][j] = dp[i-1][j] + dp[i][j-1]
        return dp[m-1][n-1]


m = 3
n = 7
result = Solution().uniquePaths(m, n)
assert result == 28

m = 3
n = 2
result = Solution().uniquePaths(m, n)
assert result == 3

m = 7
n = 3
result = Solution().uniquePaths(m, n)
assert result == 28

m = 3
n = 3
result = Solution().uniquePaths(m, n)
assert result == 6